beets/autotag/__init__.py

```diff
@@ -119,6 +119,50 @@ def _apply_metadata(
         db_obj[field] = value
 
 
+def correct_list_fields(m: LibModel) -> None:
+    """Synchronise single and list values for the list fields that we use.
+
+    That is, ensure the same value in the single field and the first element
+    in the list.
+
+    For context, the value we set as, say, ``mb_artistid`` is simply ignored:
+    Under the current :class:`MediaFile` implementation, fields ``albumtype``,
+    ``mb_artistid`` and ``mb_albumartistid`` are mapped to the first element of
+    ``albumtypes``, ``mb_artistids`` and ``mb_albumartistids`` respectively.
+
+    This means setting ``mb_artistid`` has no effect. However, beets
+    functionality still assumes that ``mb_artistid`` is independent and stores
+    its value in the database. If ``mb_artistid`` != ``mb_artistids[0]``,
+    ``beet write`` command thinks that ``mb_artistid`` is modified and tries to
+    update the field in the file. Of course nothing happens, so the same diff
+    is shown every time the command is run.
+
+    We can avoid this issue by ensuring that ``mb_artistid`` has the same value
+    as ``mb_artistids[0]``, and that's what this function does.
+
+    Note: :class:`Album` model does not have ``mb_artistids`` and
+    ``mb_albumartistids`` fields therefore we need to check for their presence.
+    """
+    if atype := m.albumtype:
+        m.albumtypes = list(dict.fromkeys([atype, *m.albumtypes]))
+    elif atypes := m.albumtypes:
+        m.albumtype = atypes[0]
+
+    if hasattr(m, "mb_artistids"):
+        if aid := m.mb_artistid:
+            m.mb_artistids = list(dict.fromkeys([aid, *m.mb_artistids]))
```

Comment:
It took me some time to understand this oneliner, but it actually was a lack of Python knowledge,. I wasn't aware of what `dict.fromkeys()` is doing. I was just trying to solve this using a dictionary comprehension but in the end, it does not get more readable:

```
m.mb_artistids = list({key: None for key in [aid, *m.mb_artistids]})
```

So I guess your solution is the most concise and best readable. Just brainstorming further, the default for all values dict.fromkeys() sets is `None`. One must know that, so when we make this explicit it would look like this:

```
m.mb_artistids = list(dict.fromkeys([aid, *m.mb_artistids], None))
```

```
m.mb_artistids = [
    dict.fromkeys([aid, *m.mb_artistids], None)
]
```

Since most of the time Python sets a default of None "for things" (eg. dict.get()) when it's not explicitly given, I guess that idea again does not make it more readable.

So yes, leave it as-is! Haha, just wanted to waste some lines of text speaking it out loud 😆

Reply:
btw, that last example does not work anyway! forget it!

Reply:
To be fair there's a lot going on in this one liner - I feel like it would be useful to find a small comment here with a summary.

See the last commit - I defined `unique_list` as a separate util function and deduped the logic that sets the fields.

Reply:
Oh wow! Yeah that reads like a book now! More words but they are english! I like it! 

Some fun fact, while we are at it. I'm not entirely sure but I recently found a helper function in the last genre plugin: 

https://github.com/beetbox/beets/blob/22163d70a77449d83670e60ad3758474463de31b/beetsplug/lastgenre/__init__.py#L48-L51

Isn't that something very similar but using `set()` to achieve the "uniqueness". Set is unordered so the comprehension takes care of that. This is off-topic here but since I plan on finally finishing some lastgenre plugin things I might make use of your utility function instead of this one. What's the difference? Thanks for the help understanding it entirely beforehand! :-)

Reply:
I suspect this function originated before 3.7 was released (before then, `dict` keys were not ordered), so `list(dict.fromkeys(...))` wasn't a thing back then. The only builtin help that was available was `list(set(...))` however this is does not keep the original order.

Generally `deduplicate` implementation is a bit of a hack and (at least personally) requires a fair bit of mental effort to read. `list(dict.fromkeys(...))` is less verbose, [more than twice as fast](https://www.peterbe.com/plog/fastest-way-to-uniquify-a-list-in-python-3.6) and more well-known to people.

Reply:
Thank you so much for shedding some light on this. So helpful! Great! :-*